// apify/build/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const utils_log_1 = require("./utils_log");
const actor_1 = require("./actor");
exports.main = actor_1.main;
exports.getEnv = actor_1.getEnv;
exports.call = actor_1.call;
exports.callTask = actor_1.callTask;
exports.metamorph = actor_1.metamorph;
exports.addWebhook = actor_1.addWebhook;
const autoscaled_pool_1 = require("./autoscaling/autoscaled_pool");
exports.AutoscaledPool = autoscaled_pool_1.default;
const basic_crawler_1 = require("./crawlers/basic_crawler");
exports.BasicCrawler = basic_crawler_1.default;
const cheerio_crawler_1 = require("./crawlers/cheerio_crawler");
exports.CheerioCrawler = cheerio_crawler_1.default;
const dataset_1 = require("./storages/dataset");
exports.pushData = dataset_1.pushData;
exports.openDataset = dataset_1.openDataset;
const events_2 = require("./events");
exports.events = events_2.default;
exports.initializeEvents = events_2.initializeEvents;
exports.stopEvents = events_2.stopEvents;
const key_value_store_1 = require("./storages/key_value_store");
exports.getValue = key_value_store_1.getValue;
exports.setValue = key_value_store_1.setValue;
exports.getInput = key_value_store_1.getInput;
exports.openKeyValueStore = key_value_store_1.openKeyValueStore;
const puppeteer_launcher_1 = require("./browser_launchers/puppeteer_launcher");
exports.launchPuppeteer = puppeteer_launcher_1.launchPuppeteer;
const playwright_launcher_1 = require("./browser_launchers/playwright_launcher");
exports.launchPlaywright = playwright_launcher_1.launchPlaywright;
const browser_crawler_1 = require("./crawlers/browser_crawler");
exports.BrowserCrawler = browser_crawler_1.default;
const puppeteer_crawler_1 = require("./crawlers/puppeteer_crawler");
exports.PuppeteerCrawler = puppeteer_crawler_1.default;
const playwright_crawler_1 = require("./crawlers/playwright_crawler");
exports.PlaywrightCrawler = playwright_crawler_1.default;
const request_1 = require("./request");
exports.Request = request_1.default;
const request_list_1 = require("./request_list");
exports.RequestList = request_list_1.RequestList;
exports.openRequestList = request_list_1.openRequestList;
const proxy_configuration_1 = require("./proxy_configuration");
exports.createProxyConfiguration = proxy_configuration_1.createProxyConfiguration;
const request_queue_1 = require("./storages/request_queue");
exports.openRequestQueue = request_queue_1.openRequestQueue;
const utils_1 = require("./utils");
exports.newClient = utils_1.newClient;
exports.getMemoryInfo = utils_1.getMemoryInfo;
exports.isAtHome = utils_1.isAtHome;
const puppeteer_utils_1 = require("./puppeteer_utils");
const playwright_utils_1 = require("./playwright_utils");
const utils_social_1 = require("./utils_social");
const enqueue_links_1 = require("./enqueue_links/enqueue_links");
const pseudo_url_1 = require("./pseudo_url");
exports.PseudoUrl = pseudo_url_1.default;
const live_view_server_1 = require("./live_view/live_view_server");
exports.LiveViewServer = live_view_server_1.default;
const utils_request_1 = require("./utils_request");
const session_pool_1 = require("./session_pool/session_pool");
exports.openSessionPool = session_pool_1.openSessionPool;
const session_1 = require("./session_pool/session");
exports.Session = session_1.Session;
// Increase the global limit for event emitter memory leak warnings.
events_1.EventEmitter.defaultMaxListeners = 50;
const exportedUtils = Object.assign(utils_1.publicUtils, {
    puppeteer: puppeteer_utils_1.puppeteerUtils,
    playwright: playwright_utils_1.playwrightUtils,
    social: utils_social_1.socialUtils,
    log: utils_log_1.default,
    enqueueLinks: enqueue_links_1.enqueueLinks,
    requestAsBrowser: utils_request_1.requestAsBrowser,
});
exports.utils = exportedUtils;
