// apify/build/storages/storage_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const consts_1 = require("apify-shared/consts");
const cache_container_1 = require("../cache_container");
const utils = require("../utils");
/* eslint-enable no-unused-vars,import/named,import/no-duplicates,import/order */
const DEFAULT_ID_ENV_VAR_NAMES = {
    Dataset: consts_1.ENV_VARS.DEFAULT_DATASET_ID,
    KeyValueStore: consts_1.ENV_VARS.DEFAULT_KEY_VALUE_STORE_ID,
    RequestQueue: consts_1.ENV_VARS.DEFAULT_REQUEST_QUEUE_ID,
};
const MAX_OPENED_STORAGES = 1000;
/**
 * @template T
 * StorageManager takes care of opening remote or local storages.
 * @property {Function} StorageConstructor
 * @property {string} name
 * @property {LruCache} cache
 * @ignore
 */
class StorageManager {
    /**
     * @param {T} StorageConstructor
     */
    constructor(StorageConstructor) {
        this.StorageConstructor = StorageConstructor;
        this.name = StorageConstructor.name;
        this.cache = cache_container_1.default.openCache(this.name, MAX_OPENED_STORAGES);
    }
    /**
     * @param {string} idOrName
     * @param {object} [options]
     * @param {boolean} [options.forceCloud]
     * @return {Promise<T>}
     */
    async openStorage(idOrName, options = {}) {
        if (!process.env[consts_1.ENV_VARS.LOCAL_STORAGE_DIR] && !process.env[consts_1.ENV_VARS.TOKEN]) {
            throw new Error(`Cannot use ${this.name} as neither ${consts_1.ENV_VARS.LOCAL_STORAGE_DIR} nor ${consts_1.ENV_VARS.TOKEN}`
                + ' environment variable is set. You need to set one these variables in order to enable data storage.');
        }
        const isLocal = !!(process.env[consts_1.ENV_VARS.LOCAL_STORAGE_DIR] && !options.forceCloud);
        if (!idOrName) {
            const defaultIdEnvVarName = DEFAULT_ID_ENV_VAR_NAMES[this.name];
            idOrName = process.env[defaultIdEnvVarName];
            if (!idOrName && isLocal)
                idOrName = consts_1.LOCAL_ENV_VARS[defaultIdEnvVarName];
            if (!idOrName)
                throw new Error(`The '${defaultIdEnvVarName}' environment variable is not defined.`);
        }
        const cacheKey = this._createCacheKey(idOrName, isLocal);
        let storage = this.cache.get(cacheKey);
        if (!storage) {
            const client = isLocal ? utils.apifyStorageLocal : utils.apifyClient;
            const storageObject = await this._getOrCreateStorage(idOrName, this.name, client);
            storage = new this.StorageConstructor({
                id: storageObject.id,
                name: storageObject.name,
                client,
                isLocal,
            });
            this._addStorageToCache(storage);
        }
        return storage;
    }
    /**
     * @param {object} storage
     * @param {string} storage.id
     * @param {string} [storage.name]
     * @param {boolean} [storage.isLocal]
     */
    closeStorage(storage) {
        const idKey = this._createCacheKey(storage.id, storage.isLocal);
        this.cache.remove(idKey);
        if (storage.name) {
            const nameKey = this._createCacheKey(storage.name, storage.isLocal);
            this.cache.remove(nameKey);
        }
    }
    /**
     * @param {string} idOrName
     * @param {boolean} isLocal
     * @return {string}
     * @ignore
     * @protected
     * @internal
     */
    _createCacheKey(idOrName, isLocal) {
        return isLocal
            ? `LOCAL:${idOrName}`
            : `REMOTE:${idOrName}`;
    }
    /**
     * Helper function that first requests storage by ID and if storage doesn't exist then gets it by name.
     * @param {string} storageIdOrName
     * @param {string} storageConstructorName
     * @param {ApifyClient|ApifyStorageLocal} apiClient
     * @ignore
     * @protected
     * @internal
     */
    async _getOrCreateStorage(storageIdOrName, storageConstructorName, apiClient) {
        const { createStorageClient, createStorageCollectionClient, } = this._getStorageClientFactories(apiClient, storageConstructorName);
        const storageClient = createStorageClient(storageIdOrName);
        const existingStorage = await storageClient.get();
        if (existingStorage)
            return existingStorage;
        const storageCollectionClient = createStorageCollectionClient();
        return storageCollectionClient.getOrCreate(storageIdOrName);
    }
    /**
     * @param {ApifyClient|ApifyStorageLocal} client
     * @param {string} storageConstructorName
     * @ignore
     * @protected
     * @internal
     */
    _getStorageClientFactories(client, storageConstructorName) {
        // Dataset => dataset
        const clientName = storageConstructorName[0].toLowerCase() + storageConstructorName.slice(1);
        // dataset => datasets
        const collectionClientName = `${clientName}s`;
        return {
            createStorageClient: client[clientName].bind(client),
            createStorageCollectionClient: client[collectionClientName].bind(client),
        };
    }
    /**
     * @param {object} storage
     * @param {string} storage.id
     * @param {string} [storage.name]
     * @param {boolean} [storage.isLocal]
     * @ignore
     * @protected
     * @internal
     */
    _addStorageToCache(storage) {
        const idKey = this._createCacheKey(storage.id, storage.isLocal);
        this.cache.add(idKey, storage);
        if (storage.name) {
            const nameKey = this._createCacheKey(storage.name, storage.isLocal);
            this.cache.add(nameKey, storage);
        }
    }
}
exports.default = StorageManager;
